src/main/java/org/jenkinsci/plugins/vsphere/VSphereCloudRetentionStrategy.java

```diff
@@ -27,7 +26,7 @@ public DescriptorImpl getDescriptor() {
         return DESCRIPTOR;
     }
 
-    @Restricted(NoExternalUse.class)
+    @Extension

```

Comment:
I am concerned that this might introduce a bug just like [this one in the docker plugin](https://github.com/jenkinsci/docker-plugin/issues/697), whose retention strategy [specifies it's an Extension](https://github.com/jenkinsci/docker-plugin/blob/docker-plugin-1.1.5/src/main/java/com/nirima/jenkins/plugins/docker/strategy/DockerOnceRetentionStrategy.java#L128) much as you have done.
The openstack-plugin's retention strategy has [this comment](https://github.com/jenkinsci/openstack-cloud-plugin/blob/openstack-cloud-2.41/src/main/java/jenkins/plugins/openstack/compute/JCloudsRetentionStrategy.java#L88) saying:
`// no @Extension since this retention strategy is used only for cloud nodes that we provision automatically.`
which makes me suspect that registering this as an Extension will cause this retention strategy to be offered to all manually-created slaves, not just vSphere ones.

Can you test this?

Reply:
Inside the descriptor, it should have a function to say which nodes this configuration is available for

Reply:
I agree that, if an external plugin is to access this class, we'll have to drop the `@Restricted(NoExternalUse.class)` bit, but are you 100% sure it has to also be annotated with `@Extension`?
While I appreciate that it probably won't get automatically discovered without that annotation, it may be possible to specify it via a full package name (I don't know for sure - I've never used JCasc).

What we need to avoid is having Jenkins think it's a general-purpose retention strategy, so if it is marked as an extension, there will have to be some form of additional filtering to prevent it from being suggested as an option for non-vSphere slaves ... and I don't know enough about how this aspect of Jenkins works to suggest what we'd have to write in the plugin code to make that happen.
(all I do know is that, if we flag it as a generic retention strategy for general use, Jenkins will make it available for generic use, and bug reports will result)

Reply:
I'll look into it, since I need this for production since tomorrow 🤭

Reply:
Well, FYI you can find a build of what you've created on the jenkinsci site, which is linked to within the "checks" shown below.
https://ci.jenkins.io/job/Plugins/job/vsphere-cloud-plugin/job/PR-100/ is the build for this PR - it has a build artifact of the hpi.

I suggest that you deploy that code for now and, when you have time, find out what's needed to enable your use-case without breaking it elsewhere - once we've got something that gives you what you need without unwanted side effects, I'd be happy to merge it in.

Reply:
I think I've found out how it'd have to work - we'd need a DescriptorVisibilityFilter.
There's a list of implementations [here](https://jenkins.io/doc/developer/extensions/jenkins-core/#descriptorvisibilityfilter).
e.g. https://github.com/KostyaSha/yet-another-docker-plugin/blob/master/yet-another-docker-plugin/src/main/java/com/github/kostyasha/yad/DockerDescriptorVisibilityFilter.java

Our implementation would have to have code which said "if the descriptor is one of our retention strategies but the context is not docker then return false" (or something like that).

Disclaimer: I'm not an expert in this; I've just read a few bits of code, e.g. [Slave.java's retentionStrategyDescriptors method](https://github.com/jenkinsci/jenkins/blob/jenkins-2.155/core/src/main/java/hudson/model/Slave.java#L623) and [DescriptorVisibilityFilter](https://github.com/jenkinsci/jenkins/blob/jenkins-2.155/core/src/main/java/hudson/model/DescriptorVisibilityFilter.java).  I've never coded one of these before, or tested how this works...

Reply:
maybe @jenkinsci/code-reviewers have an idea for this 🤭

Reply:
DescriptorVisibilityFilterImpl seems to do the the trick 
![image](https://user-images.githubusercontent.com/1661688/49907667-74dcf080-fe77-11e8-8209-a6d1c9b26a29.png)

Reply:
Hmm, looks like the Azure plugin has much the same problem that we're trying to fix here ;-)

Reply:
Definitely still need the extension, no need to remove the restricted annotation.